Upstream Kubernetes Mesos controller manager and km binary

tsenart opened the pull request:
This PR upstreams the modified Kubernetes Mesos controller manager and km binary.
It depends on #8882 and as such includes those commits too. The code that needs reviewing is in the **third and last commit**.

Closes mesosphere/kubernetes-mesos#310
Depends on GoogleCloudPlatform/kubernetes#8882

quinton-hoole commented:
Assigning to @davidopp for review, as he's already looking at the related #8882

thockin commented on cmd/integration/integration.go:
```diff
@@ -915,7 +915,7 @@ func main() {
 	go func() {
 		defer util.FlushLogs()
-		time.Sleep(3 * time.Minute)
+		time.Sleep(5 * time.Minute)
```
Explain?

jdef replied:
picked up by an earlier rebase, accidentally included as part of this PR; will remove

thockin commented on cmd/kubelet/app/server.go:
```diff
@@ -284,25 +284,10 @@ func (s *KubeletServer) Run(_ []string) error {
 		return err
 	}
 
-	if s.TLSCertFile == "" && s.TLSPrivateKeyFile == "" {
```
refactorings like this should be busted out to their own PRs

jdef replied:
will do

jdef replied:
PTAL #9315

quinton-hoole commented:
ok to test

thockin commented:
I think this raises a few really big issues.

First, this puts a bunch of new stuff in the build path.  Every build for every developer is going to slow down.  Our build has been getting slower and slower, already so I have a real probelm with this.

Second, this is a TON of code.  Do we (kubernetes) really want to carry it?  Is it really better to stick it into our repo, putting our gatekeepers between mesos devs and mesos code?  Why could this not, for example, live in a mesos project?

davidopp commented:
@thockin Could we address the first issue by excluding some or all of contrib/ from building by default?

On your second point, if the code were in the mesos repository instead of ours, what would pulling in new versions and building Kubernetes look like?

guenter commented:
Hey @thockin, the code used to live in its own repo: https://github.com/mesosphere/kubernetes-mesos
There were a couple of challenges with this:
* k8s APIs changed all the time and broke compatibility
* k8s-mesos devs had to spend days rebasing and playing catch up with k8s
* it's not a great user experience having to cobble together multiple repos to make this work on Mesos, an important platform for running k8s outside of GCP and at scale

WRT the issues you flagged
* for build time: how much slower is it? Happy to help work out a solution but given the release schedule I think this should be post-1.0
* there have been discussions around a new API for plugging in things like Mesos, but the decision was to punt on that and add Mesos support in contrib instead (see #5980)

Again, happy to address these issues after the release but we had commitment from Google committers to get this in for 1.0.

thockin commented:
I've had time to contemplate this more and I withdraw most of my objections.  It's useful to have to all together.  I don't think it should be part of the default build, though.  I'll go through and make some concrete notes.

thockin commented on build/common.sh:
```diff
@@ -311,6 +311,7 @@ function kube::build::build_image() {
     api
     build
     cmd
+    contrib/mesos
```
contrib/* should not be part of the default build.  I think it would be better to have a Makefile or something in this dir that builds everything therein.

jdef replied:
https://github.com/mesosphere/kubernetes-mesos/issues/329

thockin commented on cmd/kubelet/app/server.go:
```diff
@@ -253,7 +253,7 @@ func (s *KubeletServer) Run(_ []string) error {
 	}
 
-	client, err := s.createAPIServerClient()
+	client, _, err := s.CreateAPIServerClient()
```
This is a minor point but it feels like this could be its own PR

thockin commented on pkg/kubelet/image_manager_test.go:
```diff
@@ -216,7 +216,7 @@ func TestFreeSpaceImagesInUseContainersAreIgnored(t *testing.T) {
 	assert := assert.New(t)
 	require.NoError(t, err)
-	assert.Equal(1024, spaceFreed)
+	assert.EqualValues(1024, spaceFreed)
```
This also feels like it's own PR.  Basically, th eideal mega-PR touches *just* godeps and contrib/mesos

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -2062,7 +2062,7 @@ func getPodReadyCondition(spec *api.PodSpec, statuses []api.ContainerStatus) []a
 
 // By passing the pod directly, this method avoids pod lookup, which requires
 // grabbing a lock.
```
Comment on why this is expoorted

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -1088,7 +1088,7 @@ func (kl *Kubelet) syncPod(pod *api.Pod, mirrorPod *api.Pod, runningPod kubecont
 			// exist yet.
 			return
 		}
-		status, err := kl.generatePodStatus(pod)
```
another separate PR with explanation

tsenart commented:
Just to be sure, I want to remind that this PR depends on #8882 and as such includes those commits too. The code that needs reviewing here is in the last commit b43199b3f540dcea9dbf7e3d1dc3bae743834773.

davidopp commented:
> I want to remind that this PR depends on #8882 and as such includes those commits too. The code that needs reviewing here is in the last commit b43199b.

I reviewed just that commit yesterday and added a bunch of comments, but none of the comments showed up in this PR thread (or email thread), and now they're gone when I view that commit. Do you know where they went?

davidopp commented:
Oh never mind, it did show up in the PR, just not in the email for some reason.

jdef commented:
David, thanks for all the feedback. I've grouped this into several subtasks
for the team and we're addressing them immediately (today).

On Mon, Jun 8, 2015 at 1:35 PM, David Oppenheimer <notifications@github.com>
wrote:

> Oh never mind, it did show up in the PR, just not in the email for some
> reason.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9265#issuecomment-110085477>
> .
>

davidopp commented:
Can you please squash into one commit all of the commits that are truly part of this PR? For example, clearly 2b7dcc9 is not. The only practical way for me to review this PR is if the new stuff is in a single commit and the stuff it builds upon/depends upon (like 2b7dcc9) is not in that commit. Then I just review the one commit.

Thanks!

jdef commented:
I've squashed everything unique to this PR into a single commit:

https://github.com/mesosphere/kubernetes/commit/815f534c20909d60a9d92d77dd20d773d4fc7727

On Wed, Jun 10, 2015 at 12:55 AM, David Oppenheimer <
notifications@github.com> wrote:

> Can you please squash into one commit all of the commits that are truly
> part of this PR? For example, clearly 2b7dcc9
> <https://github.com/GoogleCloudPlatform/kubernetes/commit/2b7dcc93367385ebd72c29e65d69f9dd6ff7dded>
> is not. The only practical way for me to review this PR is if the new stuff
> is in a single commit and the stuff it builds upon/depends upon (like
> 2b7dcc9
> <https://github.com/GoogleCloudPlatform/kubernetes/commit/2b7dcc93367385ebd72c29e65d69f9dd6ff7dded>)
> is not in that commit. Then I just review the one commit.
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9265#issuecomment-110588312>
> .
>

jdef commented:
rebased to latest upstream_k8sm parent, squashed all changes for this PR into a single commit:
- https://github.com/mesosphere/kubernetes/commit/eb557a55f57f5d18881581a7c092fdbf7e4f3a88

davidopp commented:
Can you please remove from this PR everything that overlaps with 8882, and squash the rest into one commit? There are so many commits now that I'm getting lost. Thanks!

davidopp commented:
(to be specific "everything that overlaps with 8882" meaning "all the commits that are part of 8882")

Commit:
Upstream controller manager and km binary

Closes mesosphere/kubernetes-mesos#310
Depends on GoogleCloudPlatform/kubernetes#8882

- fix https://github.com/mesosphere/kubernetes-mesos/issues/336
- Fix comment typo
- Fixes https://github.com/GoogleCloudPlatform/kubernetes/pull/9265#commitcomment-11559038
- Add warning to k8s modules to also update mesos copies
- Fixes https://github.com/GoogleCloudPlatform/kubernetes/pull/9265#commitcomment-11558864
- Fixes https://github.com/GoogleCloudPlatform/kubernetes/pull/9265#commitcomment-11558855
- Add comments and TODO that hypercube and controllermanager need refactoring

jdef commented:
This comment depends on code from #8882. I've rebased this branch from `master` so that it's easier to review (the commits of #8882 are no longer visible as part of this PR).

davidopp commented:
Can you please go through all of the CAUTION comments you added and verify that these filenames are actually correct (relative to what is in #8882)? It seems that some of them are wrong.

Commit:
fix broken file refs

jdef commented:
Done

On Thu, Jun 11, 2015 at 3:20 PM, David Oppenheimer <notifications@github.com
> wrote:

> Can you please go through all of the CAUTION comments you added and verify
> that these filenames are actually correct (relative to what is in #8882
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/8882>)? It seems
> that some of them are wrong.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9265#issuecomment-111250109>
> .
>

davidopp commented:
So twice now this has failed on our Shippable

```runtime=1.4  env=KUBE_TEST_API_VERSIONS="v1",KUBE_TEST_ETCD_PREFIXES="registry"```

build but passed on our Shippable

```runtime=1.3  env=KUBE_TEST_API_VERSIONS="v1beta3",KUBE_TEST_ETCD_PREFIXES="kubernetes.io/registry" ```

build.

The failing test this time is
```
PATH=$HOME/gopath/bin:./third_party/etcd:$PATH KUBE_TEST_API_VERSIONS="${KUBE_TEST_API_VERSIONS}" KUBE_INTEGRATION_TEST_MAX_CONCURRENCY=4 LOG_LEVEL=4 ./hack/test-integration.sh
```

Can you try running it locally? End of failure log is
```
I0611 19:35:57.899741   29542 nodecontroller.go:291] Nodes ReadyCondition updated. Updating timestamp: {Capacity:map[cpu:{Amount:0 Format:DecimalSI} memory:{Amount:0 Format:DecimalSI} pods:{Amount:32.000 Format:DecimalSI}] Phase: Conditions:[{Type:Ready Status:True LastHeartbeatTime:2015-06-11 19:35:45 +0000 UTC LastTransitionTime:2015-06-11 19:35:18 +0000 UTC Reason:kubelet is posting ready status Message:}] Addresses:[{Type:LegacyHostIP Address:::1}] NodeInfo:{MachineID: SystemUUID: BootID: KernelVersion: OsImage: ContainerRuntimeVersion:docker:// KubeletVersion:v0.19.0-138-g45af3681e8be72-dirty KubeProxyVersion:v0.19.0-138-g45af3681e8be72-dirty}}
 vs {Capacity:map[cpu:{Amount:0 Format:DecimalSI} memory:{Amount:0 Format:DecimalSI} pods:{Amount:32.000 Format:DecimalSI}] Phase: Conditions:[{Type:Ready Status:True LastHeartbeatTime:2015-06-11 19:35:55 +0000 UTC LastTransitionTime:2015-06-11 19:35:18 +0000 UTC Reason:kubelet is posting ready status Message:}] Addresses:[{Type:LegacyHostIP Address:::1}] NodeInfo:{MachineID: SystemUUID: BootID: KernelVersion: OsImage: ContainerRuntimeVersion:docker:// KubeletVersion:v0.19.0-138-g45af3681e8be72-dirty KubeProxyVersion:v0.19.0-138-g45af3681e8be72-dirty}}.
I0611 19:35:57.955869   29542 config.go:244] Setting pods for source file : {[] 0 file}
F0611 19:35:58.083974   29542 integration.go:400] FAILED: pods never created timed out waiting for the condition
!!! Error in ./hack/test-integration.sh:49
  '"${KUBE_OUTPUT_HOSTBIN}/integration" --v=${LOG_LEVEL} --api-version="$1" --max-concurrency="${KUBE_INTEGRATION_TEST_MAX_CONCURRENCY}"' exited with status 255
Call stack:
  1: ./hack/test-integration.sh:49 runTests(...)
  2: ./hack/test-integration.sh:63 main(...)
Exiting with status 1
```

I will re-run Shippable again, but it's starting to look like something in this PR may be causing a problem.

jdef commented:
Strange that this PR would impact integration tests. The only core changes
are a few comments. All of the other source code lives in a directory
that's not even built. Will take a look.

On Thu, Jun 11, 2015 at 3:48 PM, David Oppenheimer <notifications@github.com
> wrote:

> So twice now this has failed on our Shippable
>
> runtime=1.4
> env=KUBE_TEST_API_VERSIONS="v1",KUBE_TEST_ETCD_PREFIXES="registry"
>
> build but passed on our Shippable
>
> runtime=1.3 env=KUBE_TEST_API_VERSIONS="v1beta3",KUBE_TEST_ETCD_PREFIXES="
> kubernetes.io/registry"
>
> build.
>
> The failing test this time is
>
> PATH=$HOME/gopath/bin:./third_party/etcd:$PATH KUBE_TEST_API_VERSIONS="${KUBE_TEST_API_VERSIONS}" KUBE_INTEGRATION_TEST_MAX_CONCURRENCY=4 LOG_LEVEL=4 ./hack/test-integration.sh
>
> Can you try running it locally? End of failure log is
>
> I0611 19:35:57.899741   29542 nodecontroller.go:291] Nodes ReadyCondition updated. Updating timestamp: {Capacity:map[cpu:{Amount:0 Format:DecimalSI} memory:{Amount:0 Format:DecimalSI} pods:{Amount:32.000 Format:DecimalSI}] Phase: Conditions:[{Type:Ready Status:True LastHeartbeatTime:2015-06-11 19:35:45 +0000 UTC LastTransitionTime:2015-06-11 19:35:18 +0000 UTC Reason:kubelet is posting ready status Message:}] Addresses:[{Type:LegacyHostIP Address:::1}] NodeInfo:{MachineID: SystemUUID: BootID: KernelVersion: OsImage: ContainerRuntimeVersion:docker:// KubeletVersion:v0.19.0-138-g45af3681e8be72-dirty KubeProxyVersion:v0.19.0-138-g45af3681e8be72-dirty}}
>  vs {Capacity:map[cpu:{Amount:0 Format:DecimalSI} memory:{Amount:0 Format:DecimalSI} pods:{Amount:32.000 Format:DecimalSI}] Phase: Conditions:[{Type:Ready Status:True LastHeartbeatTime:2015-06-11 19:35:55 +0000 UTC LastTransitionTime:2015-06-11 19:35:18 +0000 UTC Reason:kubelet is posting ready status Message:}] Addresses:[{Type:LegacyHostIP Address:::1}] NodeInfo:{MachineID: SystemUUID: BootID: KernelVersion: OsImage: ContainerRuntimeVersion:docker:// KubeletVersion:v0.19.0-138-g45af3681e8be72-dirty KubeProxyVersion:v0.19.0-138-g45af3681e8be72-dirty}}.
> I0611 19:35:57.955869   29542 config.go:244] Setting pods for source file : {[] 0 file}
> F0611 19:35:58.083974   29542 integration.go:400] FAILED: pods never created timed out waiting for the condition
> !!! Error in ./hack/test-integration.sh:49
>   '"${KUBE_OUTPUT_HOSTBIN}/integration" --v=${LOG_LEVEL} --api-version="$1" --max-concurrency="${KUBE_INTEGRATION_TEST_MAX_CONCURRENCY}"' exited with status 255
> Call stack:
>   1: ./hack/test-integration.sh:49 runTests(...)
>   2: ./hack/test-integration.sh:63 main(...)
> Exiting with status 1
>
> I will re-run Shippable again, but it's starting to look like something in
> this PR may be causing a problem.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9265#issuecomment-111257368>
> .
>

davidopp commented:
Sorry, the previous comment was supposed to go in 8882. I typed it into the wrong window. The Shippable failure here is a test infra problem. I'll re-run now.

davidopp commented:
The CLA bot is complaining that @tsenart has not signed the CLA. AFAICT he did not actually contribute code here, but to make the CLA bot happy could he sign it?

karlkfi commented:
@tsenart Is employed by Mesosphere, under our org CLA. Tho it's possible at one point he committed before adding his @mesosphere email address to github.

davidopp commented:
This looks fine, but can't test and merge until #8882 is merged.

karlkfi commented:
#8882 is merged. This is ready for testing.

davidopp commented:
This is failing on runtime=1.4, api=v1 with the following. I'll re-run shippable to see if it is a flake...

```
I0612 18:55:06.339097    5389 proc_test.go:73] do'ing deferred action
I0612 18:55:06.339566    5389 proc_test.go:77] deferred action invoked
I0612 18:55:06.352230    5389 proc_test.go:98] do'ing deferred action
I0612 18:55:06.352531    5389 proc_test.go:102] deferred action invoked
I0612 18:55:06.354076    5389 proc_test.go:124] do'ing deferred action 0
I0612 18:55:06.354208    5389 proc_test.go:124] do'ing deferred action 1
I0612 18:55:06.354264    5389 proc_test.go:124] do'ing deferred action 2
I0612 18:55:06.354326    5389 proc_test.go:124] do'ing deferred action 3
I0612 18:55:06.354400    5389 proc_test.go:124] do'ing deferred action 4
I0612 18:55:06.354497    5389 proc_test.go:124] do'ing deferred action 5
I0612 18:55:06.354756    5389 proc_test.go:124] do'ing deferred action 6
I0612 18:55:06.354817    5389 proc_test.go:124] do'ing deferred action 7
I0612 18:55:06.354884    5389 proc_test.go:124] do'ing deferred action 8
I0612 18:55:06.354951    5389 proc_test.go:124] do'ing deferred action 9
I0612 18:55:06.355364    5389 proc_test.go:128] deferred action invoked
I0612 18:55:06.355457    5389 proc_test.go:128] deferred action invoked
I0612 18:55:06.355513    5389 proc_test.go:128] deferred action invoked
I0612 18:55:06.355564    5389 proc_test.go:128] deferred action invoked
I0612 18:55:06.355605    5389 proc_test.go:128] deferred action invoked
I0612 18:55:06.355872    5389 proc_test.go:128] deferred action invoked
I0612 18:55:06.355921    5389 proc_test.go:128] deferred action invoked
I0612 18:55:06.355987    5389 proc_test.go:128] deferred action invoked
I0612 18:55:06.356040    5389 proc_test.go:128] deferred action invoked
I0612 18:55:06.356095    5389 proc_test.go:128] deferred action invoked
--- FAIL: TestProc_doWithNestedXConcurrent (9.15s)
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.438697529 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.438891261 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.4390711 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.439135224 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.439247744 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.439317746 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.43952104 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.439569051 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.439677097 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.439729592 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.439825862 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.439877167 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.439973947 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.440042382 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.44020899 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.4402693 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.440375778 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.440446603 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.440578185 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.440657819 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.440836179 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.440926366 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.441074778 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.441121318 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.441242154 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.441290526 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.441401567 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.447682099 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.447967652 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.448031602 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.448166507 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.448252115 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.448383846 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.448430661 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.448545593 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.448642213 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.448759438 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.44882802 +0000 UTC
	proc_test.go:327: runDelegationTest received executed signal at 2015-06-12 18:55:15.448934662 +0000 UTC
	proc_test.go:282: runDelegationTest finished at 2015-06-12 18:55:15.448978036 +0000 UTC
	proc_test.go:366: unexpected error: timed out waiting deferred execution
FAIL
coverage: 83.8% of statements
FAIL	github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/proc	10.157s
```

davidopp commented:
LGTM

Shippable passed.

CLA warning is apparently wrong.

davidopp commented:
Unfortunately we can't merge this until the CLA thing gets fixed. @guenter @tsenart can you guys look into why it's complaining about @tsenart?

quinton-hoole commented:
Here's what usually works for me:

"To make this automatic, please remind Mesosphere to follow the instructions that they should have received with their CLA (setting up an appropriate Google Group with members etc).

You can add your Mesosphere email address to your github account, and/or add your non-Mesosphere email address to the above Google Group."

karlkfi commented:
Created a copy to see if the bot likes me as the PR creator: #9752

davidopp commented:
Shippable is failing again with the same error as before.

davidopp commented:
Oh, and #9752 (the PR that is an identical copy of this one) is failing Shippable with exactly the same error, so I don't think this is a flake. Could the Mesosphere folks please investigate?

davidopp commented:
("Same error as before" meaning
https://github.com/GoogleCloudPlatform/kubernetes/pull/9265#issuecomment-111605132
)

davidopp commented:
LGTM

Shippable is now passing (fixed by #9775). Oncall, it is OK to merge this despite the CLA error. If you want more details, let me know and I will forward you the discussion thread we had with the CLA bot folks.

fabioy commented:
Merged. Thanks.
